File: app/src/test/java/com/example/usermanagmentapp/ui/details/UserDetailsViewModelTest.java
package com.example.usermanagmentapp.ui.details;

import static org.junit.Assert.assertEquals;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyInt;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import androidx.arch.core.executor.testing.InstantTaskExecutorRule;
import androidx.lifecycle.Observer;

import com.example.usermanagmentapp.RxImmediateSchedulerRule;
import com.example.usermanagmentapp.data.UserNotFoundError;
import com.example.usermanagmentapp.data.model.User;
import com.example.usermanagmentapp.data.repository.UserRepository;

import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TestRule;
import org.mockito.ArgumentCaptor;

import io.reactivex.rxjava3.core.Scheduler;
import io.reactivex.rxjava3.core.Single;
import io.reactivex.rxjava3.schedulers.Schedulers;

public class UserDetailsViewModelTest {

    private final UserRepository repository = mock(UserRepository.class);
    private final Scheduler testScheduler = Schedulers.trampoline();
    private final Observer<UserDetailsState> testObserver = mock(Observer.class);

    @Rule
    public TestRule rule = new InstantTaskExecutorRule();

    @Rule
    public RxImmediateSchedulerRule schedulers = new RxImmediateSchedulerRule();

    @Test
    public void testGetUserDetailsHappy() {
        User user = User.testUser;
        when(repository.user(anyInt())).thenReturn(Single.just(user));

        UserDetailsViewModel viewModel = new UserDetailsViewModel(repository, testScheduler, testScheduler);
        viewModel.state.observeForever(testObserver);
        viewModel.load(1);

        ArgumentCaptor<UserDetailsState> captor = ArgumentCaptor.forClass(UserDetailsState.class);
        verify(testObserver, times(2)).onChanged(captor.capture());
        assertEquals(UserDetailsState.Loading.class, captor.getAllValues().get(0).getClass());
        assertEquals(UserDetailsState.Success.class, captor.getAllValues().get(1).getClass());
    }

    @Test
    public void testGetUserDetailsError() {
        when(repository.user(anyInt())).thenReturn(Single.error(new Throwable()));
        UserDetailsViewModel viewModel = new UserDetailsViewModel(repository, testScheduler, testScheduler);
        viewModel.state.observeForever(testObserver);
        viewModel.load(1);

        ArgumentCaptor<UserDetailsState> captor = ArgumentCaptor.forClass(UserDetailsState.class);
        verify(testObserver, times(2)).onChanged(captor.capture());
        assertEquals(UserDetailsState.Loading.class, captor.getAllValues().get(0).getClass());
        assertEquals(UserDetailsState.Error.class, captor.getAllValues().get(1).getClass());
    }

    @Test
    public void testGetUserUserNotFound() {
        when(repository.user(anyInt())).thenReturn(Single.error(new UserNotFoundError(1)));
        UserDetailsViewModel viewModel = new UserDetailsViewModel(repository, testScheduler, testScheduler);
        viewModel.state.observeForever(testObserver);
        viewModel.load(1);

        ArgumentCaptor<UserDetailsState> captor = ArgumentCaptor.forClass(UserDetailsState.class);
        verify(testObserver, times(2)).onChanged(captor.capture());
        assertEquals(UserDetailsState.Loading.class, captor.getAllValues().get(0).getClass());
        assertEquals(UserDetailsState.UserNotFound.class, captor.getAllValues().get(1).getClass());
    }
}

File: app/src/test/java/com/example/usermanagmentapp/ui/details/UserDetailsViewModelTest.kt
package com.example.usermanagmentapp.ui.details

import androidx.arch.core.executor.testing.InstantTaskExecutorRule
import androidx.lifecycle.Observer
import com.example.usermanagmentapp.RxImmediateSchedulerRule
import com.example.usermanagmentapp.data.UserNotFoundError
import com.example.usermanagmentapp.data.model.User
import com.example.usermanagmentapp.data.repository.UserRepository
import com.example.usermanagmentapp.ui.add.AddUserState
import io.reactivex.rxjava3.core.Single
import io.reactivex.rxjava3.schedulers.Schedulers
import junit.framework.TestCase.assertEquals
import org.junit.Rule
import org.junit.Test
import org.junit.rules.TestRule
import org.mockito.kotlin.any
import org.mockito.kotlin.argumentCaptor
import org.mockito.kotlin.mock
import org.mockito.kotlin.times
import org.mockito.kotlin.verify
import org.mockito.kotlin.whenever


class UserDetailsViewModelTest {
    private val repository = mock<UserRepository>()
    private val testScheduler = Schedulers.trampoline()
    private val testObserver = mock<Observer<UserDetailsState>>()

    @get:Rule
    var rule: TestRule = InstantTaskExecutorRule()


    @get:Rule
    var schedulers = RxImmediateSchedulerRule()


    @Test
    fun testGetUserDetailsHappy() {
        val user = User.testUser
        whenever(repository.user(any())).thenReturn(Single.just(user))

        val viewModel = UserDetailsViewModel(repository, testScheduler, testScheduler)
        viewModel.state.observeForever(testObserver)
        viewModel.loadData(1)
        argumentCaptor {
            verify(testObserver, times(2)).onChanged(capture())
            assertEquals(UserDetailsState.Loading, firstValue)
            @Suppress("USELESS_IS_CHECK")
            assert(secondValue is UserDetailsState.Success)
        }
    }

    @Test
    fun testGetUserDetailsError() {
        whenever(repository.user(any())).thenReturn(Single.error(Throwable()))
        val viewModel = UserDetailsViewModel(repository, testScheduler, testScheduler)
        viewModel.state.observeForever(testObserver)
        viewModel.loadData(1)
        argumentCaptor {
            verify(testObserver, times(2)).onChanged(capture())
            assertEquals(UserDetailsState.Loading, firstValue)
            @Suppress("USELESS_IS_CHECK")
            assert(secondValue is UserDetailsState.Error)
        }
    }

    @Test
    fun testGetUserUserNotFound() {
        whenever(repository.user(any())).thenReturn(Single.error(UserNotFoundError(1)))
        val viewModel = UserDetailsViewModel(repository, testScheduler, testScheduler)
        viewModel.state.observeForever(testObserver)
        viewModel.loadData(1)
        argumentCaptor {
            verify(testObserver, times(2)).onChanged(capture())
            assertEquals(UserDetailsState.Loading, firstValue)
            @Suppress("USELESS_IS_CHECK")
            assert(secondValue is UserDetailsState.UserNotFound)
        }
    }
}